Suggestion - Two new configuration options to add #365

Aeltumn opened the pull request:
This pull request included two new configuration options: "voidExcessBits" and "requireBagSpace" which solves #365. Void excess bits simply (if enabled) stops any item entities from being spawned into the world by the feedPlayer() method. Require bag space requires the user to have a bag with available slots before being allowed to chisel.
The pull request has one known 'bug' or issue, when clicking with the chisel the message stating you don't have enough space is sent twice. I've tried to fix this but couldn't find the cause (it wasn't the obvious getHand().equals(MAIN_HAND)).
(also when the options are disabled on a client and enabled on a server voidExcessBits will work but no message is sent and requireBagSpace will not, the config could be synced on serverjoin?)

Thanks for reading/looking at my pull request, I hope it will make it into the mod.

Aeltumn committed:
#365 - Added two new configuration options.

Aeltumn committed:
Reverted eclipse changing imports in ModUtil

Aeltumn committed:
Removed random tabs eclipse left

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/AlgorithmX2/Chisels-and-Bits?pullRequest=366) <br/>All committers have signed the CLA.

AlgorithmX2 commented:
I think it maybe makes sense to maybe use a common method for the "has bag space" instead the two blocks in separate methods there are now? Unless you see a reason that doesn't work? At a glance they seemed similar enough.

We'll need to find out the reason for the double message though, have you tried setting break points to see where / how the two calls are being made? If not that would probably help identify the cause.

Besides that I think it looks alright.

Aeltumn committed:
Implemented suggested improvements

Aeltumn commented:
Making a method out of hasBagSpace() didn't occur to me earlier as the two implementations differ slightly, I changed it now though. The method hasBagSpace() is in ItemBitBag as that sounds like where it belongs. I looked into the double messages and I think it's a bug in forge where the getHand() enum is not changed correctly. For now I've added a 'cache' to the message so it never sends the message again unless you change the target BlockPos. (I've never really used breakpoints much so I didn't find anything useful using them)

AlgorithmX2 commented on src/main/java/mod/chiselsandbits/items/ItemBitBag.java:
```diff
@@ -319,4 +319,24 @@ public double getDurabilityForDisplay(
+			{
+				final ItemStack is = bp.inv.getStackInSlot( x );
+				if( ( ItemChiseledBit.getStackState( is ) == stateId && ModUtil.getStackSize( is ) < bp.inv.getInventoryStackLimit() ) || ModUtil.isEmpty( is ) ) {
+					ret = true;
```
Should just return true here to exit both loops rather then the inner most. Then change the return at the bottom to false since it failed to find space by reaching that line.

AlgorithmX2 commented:
This definitely looks a lot cleaner.

I'm not a fan of the lastPos, but I suppose its not going to break anything either; I might need to look into it some on my side to see if there is a better way to deal with it. But over all I think its almost ready to pull.

Aeltumn committed:
Removed useless ret variable

Aeltumn commented:
The ret variable was a bit of tunnel vision, previously the implementation had the return variable flipped, so it would change it to false in the for statement, this was because it needed to be true if the player had no bags (because the for would never run), so I didn't notice it. Thanks for all the comments though.

EDIT: I think I still have a bug in my code, when you use voidExcessBits but not requireBagSpace it still checks bag space when saying you’ve got a full inv, it should check inventory space instead for people who chisel single bits wihout bags. (which currently probably spams notifications)
I’ll fix it as soon as I get out of school.

Aeltumn committed:
Fixed minor bugs

Aeltumn commented:
I've been trying to fix some more minor bugs and I've run into something, when trying to chisel a chiseled block ModUtil.getStateId() doesn't return the correct stateId anymore. I couldn't figure out how to find out exactly what part of the chiseled block the player targets, then check what stateId that part has so I opted to make the message say it requires 'empty slots' instead of 'empty space' and leave the code as it is.

AlgorithmX2 commented:
At a glance maybe it makes sense to check for room here?

https://github.com/MineDaveXD/Chisels-and-Bits/blob/9af4f2db9aebacd8ea8c449be102f490c62198ca/src/main/java/mod/chiselsandbits/items/ItemChisel.java#L336

and either continue if voiding is enabled, or stop if there is no room? Might be a little slow though for large edits... Hmm...

The voiding should work rather well, but the space check probably is a bit difficult to be accurate and performance friendly.

It more or less is working for the intended purpose at this point right? We might want to add a note that voiding bits can/will prevent Undo from working as it will delete the required materials. But beside that it might work fine for now.

Aeltumn committed:
Fixed voidExcessBits with negative print

Aeltumn commented:
I've added the 'bag is full' messages to the negative print, I forgot to test those earlier. I do agree checking for space is going to get bad for performance, so this should do for now. The setting is intented for players who have the stack sizes pretty high and just make new bags when it says 'no room' either way.

I think I've fixed everything for now, I'm going to be playing on the survival server (running these custom config options if they're finished in time), which Noxite (reporter of #365 ) mentioned in his report, along with 100 other players. So if any errors occur I can find them easily I'll try to fix those.
